### PR TITLE
Map ClickHouse UUID to Trino UUID #7097

### PR DESCRIPTION
**Pull request by wgzhao**
1. Add support for querying ClickHouse table with UUID data type
2. Add support for creating ClickHouse table with Trino uuid data type on Trino server side.

Fixes https://github.com/trinodb/trino/issues/7097

### CODE REVIEW

**Comment by wgzhao** [comment]
use case: 

## create table on clickhouse side

```sql
clickhouse:) create table t_uuid(id UUID) ENGINE=Log;

CREATE TABLE t_uuid
(
    `id` UUID
)
ENGINE = Log

Ok.
```

First, insert some data on clickhouse side

```sql
clickhouse :) insert into t_uuid select generateUUIDv4();

INSERT INTO t_uuid SELECT generateUUIDv4()

Ok.

0 rows in set. Elapsed: 0.001 sec.

clickhouse :) insert into t_uuid select generateUUIDv4();

INSERT INTO t_uuid SELECT generateUUIDv4()

Ok.

clickhouse :) select * from t_uuid;

SELECT *
FROM t_uuid

┌───────────────────────────────────id─┐
│ 58b6bdd6-e93e-4dec-b004-e9b55b0abd6d │
│ 33467c49-8c0a-4e3b-8513-a88cd3dc5ab6 │
└──────────────────────────────────────┘

2 rows in set. Elapsed: 0.001 sec.
```

then, query the table and insert data on Trino side 

```sql
trino:test> desc t_uuid;
 Column | Type | Extra | Comment
--------+------+-------+---------
 id     | uuid |       |
(1 row)

trino:test> select * from t_uuid;
                  id
--------------------------------------
 58b6bdd6-e93e-4dec-b004-e9b55b0abd6d
 33467c49-8c0a-4e3b-8513-a88cd3dc5ab6
(2 rows)

trino:test> insert into t_uuid select uuid();
INSERT: 1 row

trino:test> select * from t_uuid;
                  id
--------------------------------------
 58b6bdd6-e93e-4dec-b004-e9b55b0abd6d
 33467c49-8c0a-4e3b-8513-a88cd3dc5ab6
 ad428031-403b-b075-eb74-f2f289104fa2
(3 rows)
```

## create table on Trino side

```sql

trino:test> create table clickhouse.test.trino_t_uuid(uid uuid) with (ENGINE='Log');
CREATE TABLE
trino:test> insert into clickhouse.test.trino_t_uuid select uuid();
INSERT: 1 row

Query 20210514_054816_00001_g2gex, FINISHED, 1 node
Splits: 35 total, 35 done (100.00%)
1.56 [0 rows, 0B] [0 rows/s, 0B/s]

trino:test> insert into clickhouse.test.trino_t_uuid select uuid();
INSERT: 1 row

Query 20210514_054819_00002_g2gex, FINISHED, 1 node
Splits: 35 total, 35 done (100.00%)
0.92 [0 rows, 0B] [0 rows/s, 0B/s]

trino:test> select * from clickhouse.test.trino_t_uuid;
                 uid
--------------------------------------
 bc4493cf-a7a1-7840-2c36-2bf889f9eb95
 fe44cdfe-93ff-be87-8820-01590e9d5282
(2 rows)

Query 20210514_054827_00003_g2gex, FINISHED, 1 node
Splits: 17 total, 17 done (100.00%)
0.71 [2 rows, 0B] [2 rows/s, 0B/s]
```

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java [hunk 1]
```diff
@@ -109,23 +115,23 @@
 import static java.lang.String.join;
 
 public class ClickHouseClient
-        extends BaseJdbcClient
-{
+        extends BaseJdbcClient {
```
Comment (findepi) [inline plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java:118]: Can you please reformat the code with the project's code style?
see https://github.com/trinodb/trino/blob/master/DEVELOPMENT.md#code-style for how to configure it

Comment (wgzhao) [reply plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java:118]: Ok, I configure it

**Review by findepi — APPROVED** [review]
looks good! 

just editorial comments

File: plugin/trino-clickhouse/pom.xml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
```
[the remark this answers is not in the document]
Comment (wgzhao) [reply plugin/trino-clickhouse/pom.xml:2]: OK , I will revert it

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java [hunk 3]
```diff
@@ -142,18 +149,35 @@ public ClickHouseClient(
...
+    }
+
     @Override
     public Optional<JdbcExpression> implementAggregation(ConnectorSession session, AggregateFunction aggregate, Map<String, ColumnHandle> assignments)
     {
         // TODO support complex ConnectorExpressions
         return aggregateFunctionRewriter.rewrite(session, aggregate, assignments);
     }
 
-    private static Optional<JdbcTypeHandle> toTypeHandle(DecimalType decimalType)
```
[the remark this answers is not in the document]
Comment (wgzhao) [reply plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java:152]: What kind of sorting method do we want to use?

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java [hunk 4]
```diff
@@ -459,4 +486,12 @@ else if (prop.size() == 1) {
             return Optional.of("(" + String.join(",", prop) + ")");
         }
     }
+
+    private ColumnMapping uuidColumnMapping()
```
[the remark this answers is not in the document]
Comment (wgzhao) [reply plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java:490]: ok

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java [hunk 5]
```diff
@@ -459,4 +465,35 @@ else if (prop.size() == 1) {
...
+            long low = Long.reverseBytes(value.getLong(SIZE_OF_LONG));
+            UUID uuid = new UUID(high, low);
+            statement.setObject(index, uuid, Types.OTHER);
+        };
+    }
+
+    private static Slice uuidSlice(UUID uuid)
+    {
+        return wrappedLongArray(
+                Long.reverseBytes(uuid.getMostSignificantBits()),
```
Comment (hashhar) [inline plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java:496]: nit: static import `reverseBytes`

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClientModule.java [hunk 6]
```diff
@@ -33,6 +33,14 @@
 public class ClickHouseClientModule
         implements Module
 {
+    @Provides
```
Comment (hashhar) [inline plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClientModule.java:36]: Revert unrelated change.
Ideally we'd want the entire codebase to use formatter but that causes git history difficult to use.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseTypeMapping.java [hunk 7]
```diff
@@ -320,11 +321,11 @@ public void testEnum()
...
-                // TODO map to Trino UUID
-                .addRoundTrip("UUID", "'417ddc5d-e556-4d27-95dd-a34d84e46a50'", createUnboundedVarcharType(), "CAST('417ddc5d-e556-4d27-95dd-a34d84e46a50' AS varchar)")
-                .execute(getQueryRunner(), clickhouseCreateAndInsert("tpch.test_uuid"));
-
-        // TODO add test with UUID written from Trino
+                .addRoundTrip("UUID", "'114514ea-0601-1981-1142-e9b55b0abd6d'", UuidType.UUID, "cast('114514ea-0601-1981-1142-e9b55b0abd6d' as UUID)")
+                .execute(getQueryRunner(), clickhouseCreateAndInsert("default.ck_test_uuid"));
+        SqlDataTypeTest.create()
+                .addRoundTrip("uuid", "'114514ea-0601-1981-1142-e9b55b0abd6d'", UuidType.UUID, "cast('114514ea-0601-1981-1142-e9b55b0abd6d' as uuid)")
+                .execute(getQueryRunner(), trinoCreateAsSelect("default.ck_test_uuid"));
```
Comment (hashhar) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseTypeMapping.java:328]: We should also test with CREATE + INSERT from Trino. That follows a different code path.

Comment (hashhar) [reply plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseTypeMapping.java:328]: Let's also test NULL values.

And it looks to me that the arguments passed to `addRoundTrip` can be reused for both cases (the only difference is lowercase vs uppercase UUID)?

**Review by hashhar — COMMENTED** [review]
Thanks.

Mostly editorials. Some tests around NULL values would be useful.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseTypeMapping.java [hunk 8]
```diff
@@ -320,11 +323,25 @@ public void testEnum()
...
+                .execute(getQueryRunner(), trinoCreateAndInsert("default.ck_test_uuid"));
+
+        // test null value
+        String tableName = "test_uuid_null" + randomTableSuffix();
+        assertUpdate("CREATE TABLE " + tableName + "(x uuid)");
+        assertTrue(getQueryRunner().tableExists(getSession(), tableName));
+        assertUpdate("INSERT INTO " + tableName + " VALUES(null)", 1);
+        assertQuery("SELECT COUNT(*) FROM " + tableName, " VALUES (1)");
+        assertQuery("SELECT COUNT(*) FROM " + tableName + " WHERE x IS null", " VALUES (1)");
+        assertUpdate("DROP TABLE " + tableName);
```
Comment (hashhar) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseTypeMapping.java:344]: You can simply do this instead. If the types are explicitly provided in the `addRoundTrip` then they are used instead of doing a CREATE TABLE AS SELECT.

```java
@Test
public void testUuid()
{
    SqlDataTypeTest.create()
            .addRoundTrip("Nullable(UUID)", "NULL", UuidType.UUID, "CAST(NULL AS UUID)")
            .addRoundTrip("Nullable(UUID)", "'114514ea-0601-1981-1142-e9b55b0abd6d'", UuidType.UUID, "CAST('114514ea-0601-1981-1142-e9b55b0abd6d' AS UUID)")
            .execute(getQueryRunner(), clickhouseCreateAndInsert("default.ck_test_uuid"));
                                                                                                                                                            
    SqlDataTypeTest.create()
            .addRoundTrip("CAST(NULL AS UUID)", "cast(NULL as UUID)")
            .addRoundTrip("UUID '114514ea-0601-1981-1142-e9b55b0abd6d'", "CAST('114514ea-0601-1981-1142-e9b55b0abd6d' AS UUID)")
            .execute(getQueryRunner(), trinoCreateAsSelect("default.ck_test_uuid"))
            .execute(getQueryRunner(), trinoCreateAndInsert("default.ck_test_uuid"));
}
```

Comment (wgzhao) [reply plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseTypeMapping.java:344]: got it

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java [hunk 9]
```diff
@@ -149,11 +158,6 @@ public ClickHouseClient(
         return aggregateFunctionRewriter.rewrite(session, aggregate, assignments);
     }
 
-    private static Optional<JdbcTypeHandle> toTypeHandle(DecimalType decimalType)
```
[the remark this answers is not in the document]
Comment (wgzhao) [reply plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java:152]: OK, I will revert

**Review by hashhar — APPROVED** [review]

**Comment by hashhar** [comment]
@wgzhao Please address the CI failure (clickhouse test failed). LGTM after that.

**Comment by wgzhao** [comment]
@hashhar thanks
how can i re-run CI without commit ?

**Comment by hashhar** [comment]
@wgzhao You can push an empty commit like `git commit  --allow-empty -m "empty to re-run ci"`. They get removed when merging a PR because we use `Rebase and merge`.

Another option is to amend your previous commit and force-push but that makes it difficult for reviewer to make sure if anything changed in the force-push.

**Comment by wgzhao** [comment]
Thanks a lot
